llama-index-core/llama_index/core/query_pipeline/query.py

```diff
@@ -511,7 +572,9 @@ def _process_component_output(
 
         return new_queue
 
-    def _run_multi(self, module_input_dict: Dict[str, Any]) -> Dict[str, Any]:
+    def _run_multi(
+        self, module_input_dict: Dict[str, Any]
+    ) -> Tuple[Dict[str, Any], Dict[str, Dict[str, Any]]]:
```

Comment:
Hmm, suggestion: rather than changing the output types (which is technically a breaking change here), should we add a new method, something like `pipeline.run_with_intermediate_outputs(query="query")` ?

Reply:
I'm also not sold on that name (but I also can't think of a better name haha)

Reply:
Sure, would it be too much boilerplate code though?

Reply:
I think under the hood, we could have a function that probably always returns intermediate outputs, and then `run()` and `run_with_intermediate_outputs()` will return the expected output

This would minimize any code duplication

Reply:
If I'm understanding correctly, we'd like: 

`run()` and `run_with_intermediate_outputs()` both calling: `_run()` which takes in extra parameter `show_intermediate_outputs`

then `_run()` calling `_run_multi()`?

or do we want:

`_run()` and `_run_with_intermediate_outputs()` calling `_run_multi()` and `_run_multi_with_intermediate_outputs()` respectively?

The latter seems a bit repetitive to me (6 functions)

Reply:
Yea I think my thoughts was something along the first

Hides the complexity from the user while giving them pretty deterministic/expected outputs

Reply:
Ok I have implemented the `show intermediate output` feature by making these new functions:

- `run_with_intermediate_outputs()`
- `run_multi_with_intermediate_outputs()`
- `arun_with_intermediate_outputs()`
- `arun_multi_with_intermediate_outputs()`

I'm not sure I understand why we are taking this approach instead of adding `show_intermediate_outputs` as a parameter. When we add `batch_mode` in the future, wouldn't we have to add parameters anyway?

Reply:
Tbh I'm just trying to avoid a breaking change lol 

Changing the output of run() to a tuple (and a tuple that's only populated if some class var is set) feels kind of janky?

I'll take a look at this 👍

Reply:
Oh, for that, I changed `_run_multi()` to only return a tuple if `show_intermediate_output()` is `True`. Otherwise, return just the output itself w/out the tuple. So the client facing `run()` would still return just the plain output if `show_intermediate_output()=False`

Reply:
Fair enough. But having an interface that changes output types feels, inconsistent?

This is something we've tried to move away from doing. For example, agents have chat and stream_chat, rather than having a single method that can optionally do both 

Hopefully that makes sense 😅

Reply:
Ok that makes sense, thanks!